### PR TITLE
Add opaque pointers

### PR DESCRIPTION
**Pull request by xlambein**
## Description

All changes apply to LLVM >= 15.0.

- Add a `ptr_type` method to `Context` to get an opaque pointer type.
- Deprecate the use of `ptr_type` on `*Type`.
- Add a `is_opaque` method to `PointerType` to check whether the pointer is opaque (which is always true unless opaque-by-default is disabled).
- Change the tests and examples to use the new `ptr_type` method on `Context`.

Note that I haven't found a way in the C API (at least as exposed by `llvm_sys`) to disable the opaque-by-default behaviour. Because of that, I'm unsure it even makes sense to have the `is_opaque` method. Because of this I haven't really documented it much, but I can add an example to the doc if needed.

## Related Issue

#421 

## How This Has Been Tested

- Tested on 14.0, 15.0 and 17.0.
- I've added a test for `is_opaque`, but in effect the function always returns `true`.

## Breaking Changes

None yet. In the future we might want to remove the `ptr_type` method on `*Type` for LLVM 17 onward, since only opaque pointers are supported.

## Checklist

- [x] I have read the [Contributing Guide](https://github.com/TheDan64/inkwell/blob/master/.github/CONTRIBUTING.md)

### CODE REVIEW

**Comment by TheDan64** [comment]
Are ptrs always opaque in 16(?)+? (the version after whenever they were introduced?)

**Comment by xlambein** [comment]
> Aren't ptrs always opaque in 16(?)+? (the version after whenever they were introduced?)
> 
> One of your tests seems to confirm this since `context.i32_type().ptr_type(AddressSpace::default()).is_opaque()` always passes in 15+ 🤔

I think this section answers those questions: https://llvm.org/docs/OpaquePointers.html#temporarily-disabling-opaque-pointers

Basically, pointers are opaque by default starting in 15+, but can be disabled, at least in `clang`. I couldn't find a way to disable them in LLVM's C API, but tbf I didn't look very hard. Given that, I think I could remove the `is_opaque` method.

**Comment by TheDan64** [comment]
@xlambein could you please rebase this on master? Thanks

**Comment by nsumner** [comment]
Not sure if it is relevant to this specific PR, but I think getting the value type of GlobalValues also matters now with opaque pointers. I just started playing with Inkwell, though, so there may be other idioms I am missing.

e.g.
```rust
fn get_value_type<'ctx>(global: GlobalValue<'ctx>) -> AnyTypeEnum<'ctx> {
    unsafe {
        AnyTypeEnum::new(llvm_sys::core::LLVMGlobalGetValueType(
            global.as_value_ref(),
        ))
    }
}
```

**Comment by xlambein** [comment]
I'm on holiday until late this month, so I'll finish this PR then :-)

@nsumner That seems right, I'll take a closer look later and perhaps add it to the PR.

**Comment by xlambein** [comment]
Rebased and added `get_value_type` to the PR.

File: src/context.rs [hunk 1]
```diff
@@ -917,6 +924,26 @@ impl Context {
         self.context.ppc_f128_type()
     }
 
+    /// Gets the `IntType` representing 8 bit width. It will be assigned the current context.
```
Comment (gavrilikhin-d) [inline src/context.rs:927]: Description is invalid

File: src/context.rs [hunk 2]
```diff
@@ -1765,6 +1792,26 @@ impl<'ctx> ContextRef<'ctx> {
...
```
Comment (gavrilikhin-d) [inline src/context.rs:1795]: Description is invalid here too

**Comment by TheDan64** [comment]
Apologies for this taking a while, I am traveling this month and won't have much time to review until April

**Comment by TheDan64** [comment]
Not sure why the build is failing suddenly...

**Comment by TheDan64** [comment]
Ah, I tried updating the PR and it seems I broke something - please take a look and then we can get this merged

**Comment by xlambein** [comment]
@TheDan64 I've rebased and tested locally, seems to work again.

**Comment by TheDan64** [comment]
Looks like this needs to be adjusted to handle other versions

**Comment by xlambein** [comment]
I've fixed that issue :-)

**Comment by gavrilikhin-d** [comment]
@TheDan64 can you have a look?

**Comment by TheDan64** [comment]
This looks pretty good, but needs to be updated for llvm18

**Comment by xlambein** [comment]
@TheDan64 Rebased + added the necessary changes for LLVM 18. At the same time, there was a failing doctest for 18 in upstream/master that I fixed (in `src/targets.rs`), I assume because there isn't any CI for LLVM 18 yet.

**Review by TheDan64 — APPROVED** [review]
Thanks!

**Pull request merged**